packages/core/src/recoverable_state.ts

```diff
@@ -0,0 +1,220 @@
+import { Configuration, EventType, KeyValueStore } from '@crawlee/core';
+
+import type { Log } from '@apify/log';
+import log from '@apify/log';
+
+export interface RecoverableStatePersistenceOptions {
+    /**
+     * The key under which the state is stored in the KeyValueStore
+     */
+    persistStateKey: string;
+
+    /**
+     * Flag to enable or disable state persistence
+     */
+    persistenceEnabled?: boolean;
+
+    /**
+     * The name of the KeyValueStore to use for persistence.
+     * If neither a name nor an id are supplied, the default store will be used.
+     */
+    persistStateKvsName?: string;
+
+    /**
+     * The identifier of the KeyValueStore to use for persistence.
+     * If neither a name nor an id are supplied, the default store will be used.
+     */
+    persistStateKvsId?: string;
+}
+
+/**
+ * Options for configuring the RecoverableState
+ */
+export interface RecoverableStateOptions<TStateModel = Record<string, unknown>>
+    extends RecoverableStatePersistenceOptions {
+    /**
+     * The default state used if no persisted state is found.
+     * A deep copy is made each time the state is used.
+     */
+    defaultState: TStateModel;
+
+    /**
+     * A logger instance for logging operations related to state persistence
+     */
+    logger?: Log;
+
+    /**
+     * Configuration instance to use
+     */
+    config?: Configuration;
+
+    /**
+     * Optional function to transform the state to a JSON string before persistence.
+     * If not provided, JSON.stringify will be used.
+     */
+    serialize?: (state: TStateModel) => string;
+
+    /**
+     * Optional function to transform a JSON-serialized object back to the state model.
+     * If not provided, JSON.parse is used.
+     * It is advisable to perform validation in this function and to throw an exception if it fails.
+     */
+    deserialize?: (serializedState: string) => TStateModel;
+}
+
+/**
+ * A class for managing persistent recoverable state using a plain JavaScript object.
+ *
+ * This class facilitates state persistence to a `KeyValueStore`, allowing data to be saved and retrieved
+ * across migrations or restarts. It manages the loading, saving, and resetting of state data,
+ * with optional persistence capabilities.
+ *
+ * The state is represented by a plain JavaScript object that can be serialized to and deserialized from JSON.
```

Comment:
I suppose this is no longer completely true with custom `serialize` and `deserialize` methods, but there is IMO nothing wrong with not telling the users the whole truth, keeping them on the safe side.

Reply:
Well, "plain JavaScript object" is not exactly a legal term and we do accept stuff that can be (de)serialized from/to JSON, if only with the caveat that sometimes you have to write the serialization logic by hand.

![image](https://github.com/user-attachments/assets/dacfb318-046a-4d5a-a06f-3ea9dba6048b)